### PR TITLE
utils_test/libvirt: API to check migration triggered and vm state

### PR DESCRIPTION
**Pull request by balamuruhans**
For executing commands during migration of guests, ensure the migration is really started and execute it. supporting APIs are included as part of this patch.

### CODE REVIEW

File: virttest/utils_test/libvirt.py [hunk 1]
```diff
@@ -1333,17 +1333,25 @@ def do_migration(self, vms, srcuri, desturi, migration_type,
...
                 migration_thread.start()
                 eclipse_time = 0
+                stime = int(time.time())
                 if func:
-                    stime = int(time.time())
-                    if func == process.run:
-                        try:
-                            func(args['func_params'], shell=args['shell'])
-                        except KeyError:
+                    # we execute command once the migration is started
```
Comment (chunfuwen) [inline virttest/utils_test/libvirt.py]: Better remove 'we' this word

File: virttest/utils_test/libvirt.py [hunk 2]
```diff
@@ -1333,17 +1333,25 @@ def do_migration(self, vms, srcuri, desturi, migration_type,
...
+                    if self.check_migrate_started(vm):
+                        if func == process.run:
+                            logging.info("Migration started for %s", vm.name)
+                            try:
```
Comment (chunfuwen) [inline virttest/utils_test/libvirt.py]: maybe move try ... except to outer space to just under  check_migrate_started(vm)

File: virttest/utils_test/libvirt.py [hunk 3]
```diff
@@ -1333,17 +1333,25 @@ def do_migration(self, vms, srcuri, desturi, migration_type,
...
+                        else:
                             func(args['func_params'])
                     else:
-                        func(args['func_params'])
-                    eclipse_time = int(time.time()) - stime
-                    logging.debug("start_time:%s, eclipse_time:%s", stime, eclipse_time)
+                        logging.error("Migration failed to start for %s",
+                                      vm.name)
+                eclipse_time = int(time.time()) - stime
+                logging.debug("start_time:%s, eclipse_time:%s", stime, eclipse_time)
```
Comment (chunfuwen) [inline virttest/utils_test/libvirt.py]: stime is int, so %d is more suitable.

File: virttest/utils_test/libvirt.py [hunk 4]
```diff
@@ -1410,6 +1418,39 @@ def cleanup_dest_vm(self, vm, srcuri, desturi):
...
+        :return: True if state of VM is as expected, False otherwise
+        """
+        if not virsh.domain_exists(vm.name, uri=uri):
+            return False
+        vm_state = virsh.domstate(vm.name, uri=uri).stdout.strip()
+        return vm_state.lower() == state.lower()
+
+    def check_migrate_started(self, vm, state='paused', uri=None, timeout=60):
+        """
+        checks whether migration is started of not
```
Comment (chunfuwen) [inline virttest/utils_test/libvirt.py]: of to or

File: virttest/utils_test/libvirt.py [hunk 5]
```diff
@@ -1410,6 +1418,39 @@ def cleanup_dest_vm(self, vm, srcuri, desturi):
...
+        :param timeout: time in seconds to wait for migration to start
+
+        :return: True if migration is started False otherwise
+        """
+        def check_state():
+            try:
+                return self.check_vm_state(vm, state, uri)
+            except Exception:
+                return False
+        return utils_misc.wait_for(check_state, timeout)
```
Comment (chunfuwen) [inline virttest/utils_test/libvirt.py]: can we just use check_vm_state() directly, rather than wrap one duplicate func?

Comment (balamuruhans) [reply virttest/utils_test/libvirt.py]: we have to pass few parameters to the check_vm_state(), so we use duplicate func given input to wait_for()

File: virttest/utils_test/libvirt.py (hunk 2 above)
Comment (balamuruhans) [reply virttest/utils_test/libvirt.py]: it is better to be in same place, as for process.run user may or may not give shell=True, it was meant to handle that.

**Comment by balamuruhans** [comment]
@chunfuwen Thanks for reviewing, made the changes as suggested.

**Comment by balamuruhans** [comment]
@sathnaga made the changes as suggested.

**Comment by balamuruhans** [comment]
Thanks @sathnaga @chunfuwen, let us merge it.

**Pull request merged**